src/canister_tests/src/tests.rs

```diff
@@ -66,6 +66,246 @@ fn registration_with_mismatched_sender_fails() {
+            .iter()
+            .any(|device| device == &device_data_1()));
+        assert!(anchor_info
+            .devices
+            .iter()
+            .any(|device| device == &device_data_2()));
+        Ok(())
+    }
+
+    /// Verifies that a device can be removed.
+    #[test]
+    fn should_remove_device() -> Result<(), CallError> {
+        let env = StateMachine::new();
+        let canister_id = framework::install_ii_canister(&env, framework::II_WASM.clone());
+        let user_number = flows::register_anchor(&env, canister_id);
+
+        api::add(
+            &env,
+            canister_id,
+            principal_1(),
+            user_number,
+            device_data_2(),
+        )?;
+        let devices = api::lookup(&env, canister_id, user_number)?;
+        assert!(devices.iter().any(|device| device == &device_data_2()));
+
+        api::remove(
+            &env,
+            canister_id,
+            principal_1(),
+            user_number,
+            device_data_2().pubkey,
+        )?;
+
+        let devices = api::lookup(&env, canister_id, user_number)?;
+        assert_eq!(devices.len(), 1);
+        assert!(!devices.iter().any(|device| device == &device_data_2()));
+        Ok(())
+    }
+
+    /// Verifies that the even last device can be removed.
+    /// This behaviour should be changed because it makes anchors unusable, see L2-745.
+    #[test]
+    fn should_remove_last_device() -> Result<(), CallError> {
+        let env = StateMachine::new();
+        let canister_id = framework::install_ii_canister(&env, framework::II_WASM.clone());
+        let user_number = flows::register_anchor(&env, canister_id);
+
+        api::remove(
+            &env,
+            canister_id,
+            principal_1(),
+            user_number,
+            device_data_1().pubkey,
+        )?;
+
+        let devices = api::lookup(&env, canister_id, user_number)?;
+        assert!(devices.is_empty());
+        Ok(())
+    }
+
+    /// Verifies that users can only remove their own devices.
+    #[test]
+    fn should_not_remove_device_of_different_user() {
+        let env = StateMachine::new();
+        let canister_id = framework::install_ii_canister(&env, framework::II_WASM.clone());
+        flows::register_anchor_with(&env, canister_id, principal_1(), device_data_1());
+        let user_number_2 =
+            flows::register_anchor_with(&env, canister_id, principal_2(), device_data_2());
+
+        let result = api::remove(
+            &env,
+            canister_id,
+            principal_1(),
+            user_number_2,
+            device_data_2().pubkey,
+        );
+
+        expect_user_error_with_message(
+            result,
+            CanisterCalledTrap,
+            Regex::new("[a-z\\d-]+ could not be authenticated.").unwrap(),
+        );
+    }
+
+    /// Verifies that the same device cannot be added twice.
+    #[test]
+    fn should_not_add_existing_device() -> Result<(), CallError> {
```

Comment:
nit: how about having all the `add` tests together? otherwise going back and forth between remove and add is a bit confusing